internal/rotate_est.go

```diff
@@ -0,0 +1,207 @@
+package internal
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+
+	"github.com/ThalesIgnite/crypto11"
+	"github.com/miekg/pkcs11"
+	"go.mozilla.org/pkcs7"
+)
+
+var (
+	oidKeyUsage         = asn1.ObjectIdentifier([]int{2, 5, 29, 15})
+	oidExtendedKeyUsage = asn1.ObjectIdentifier([]int{2, 5, 29, 37})
+
+	asn1DigitalSignature = []byte{3, 2, 7, 128}
+	asn1TlsWebClientAuth = []byte{48, 10, 6, 8, 43, 6, 1, 5, 5, 7, 3, 2}
+)
+
+type estStep struct{}
+
+func (s estStep) Name() string {
+	return "Generate new certificate"
+}
+
+func (s estStep) Execute(handler *CertRotationHandler) error {
+	// Find the current certificate so we can build the proper EST request
+	tlsCert := handler.client.Transport.(*http.Transport).TLSClientConfig.Certificates[0]
+	cert, err := x509.ParseCertificate(tlsCert.Certificate[0])
+	if err != nil {
+		return nil
+	}
+
+	var signer crypto.Signer
+	var newKey string
+
+	// Generate a new private key
+	if handler.usePkcs11() {
+		newKey = s.nextPkeyId(handler)
+		pubAttr, err := crypto11.NewAttributeSetWithIDAndLabel(idToBytes(newKey), []byte("tls"))
+		if err != nil {
+			return fmt.Errorf("Unable to define pkcs11 attributes for new key: %w", err)
+		}
+		// The default ecdsa logic in crypto11 does not include the ability to
+		// derive which is required for ECIES decryption
+		pubAttr.AddIfNotPresent([]*pkcs11.Attribute{pkcs11.NewAttribute(pkcs11.CKA_DERIVE, true)})
+		privAttr := pubAttr.Copy()
+		signer, err = handler.crypto.ctx.GenerateECDSAKeyPairWithAttributes(pubAttr, privAttr, elliptic.P256())
+		if err != nil {
+			return fmt.Errorf("Unable to generate new keypair in HSM: %w", err)
+		}
+	} else {
+		key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return fmt.Errorf("Unable to generate new private key: %w", err)
+		}
+		keyBytes, err := x509.MarshalECPrivateKey(key)
+		if err != nil {
+			return fmt.Errorf("Unable to serialize new private key: %w", err)
+		}
+		keyBytes = pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: keyBytes})
+		newKey = string(keyBytes)
+		signer = key
+	}
+
+	// Ask EST server for new cert
+	csrBytes, err := createB64CsrDer(signer, cert)
+	if err != nil {
+		return err
+	}
+
+	url := handler.State.EstServer + "/simplereenroll"
+	res, err := handler.client.Post(url, "application/pkcs10", bytes.NewBuffer(csrBytes))
```

Comment:
If the post fails due to absence of network connectivity how exactly the retry happens? I suppose the fioconfig systemd service keeps running `fioconfig daemon` which is supposed to call `fioconfig renew-cert` on each iteration until the cert renew succeeds?

Reply:
The cover letter of the PR notes some follow up PRs required to complete this task. This is one of the items.

Reply:
I am not saying that something is missing :), just asking the questions in order to make sure I fully understand the implementation.